### PR TITLE
Show diff of changes on config push

### PR DESCRIPTION
**Pull request by gonzaloriestra**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/internal-cli-foundations/issues/765

### WHAT is this pull request doing?

- Shows a confirmation prompt to confirm the changes. It ignores comments and ordering changes.
diff
- Does nothing if there are no changes
no-changes
- Skips the confirmation when passing `--force` or `-f`

### How to test your changes?

Modify the local configuration and run `p shopify app config push`

### Measuring impact

How do we know this change was effective? Please choose one:

- [x] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [x] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

**Comment by ryancbahan** [comment]
awesome! question for future state: would it be more readable to show the diff in something more like a table view, with `prev` and `current` headers or something similar?

**Comment by gonzaloriestra** [comment]
> awesome! question for future state: would it be more readable to show the diff in something more like a table view, with prev and current headers or something similar?

@ryancbahan that's something we can explore. I also find clearer two column diffs in general, but we have to take into account that the width of the terminal (or the number of chars we can print) is usually much smaller than a browser, so it's probably hard to fit it.

File: packages/app/src/cli/prompts/config.ts [hunk 1]
```diff
@@ -64,3 +71,27 @@ export function validate(value: string): string | undefined {
   // Max filename size for Windows/Mac including the prefix/postfix
   if (result.length > 238) return 'The file name is too long.'
 }
+
+export async function confirmPushChanges(options: PushOptions, app: App) {
+  if (options.force) return true
+
+  const {configuration, configurationPath} = options
+  const localConfiguration = await readFile(configurationPath)
```
[the remark this answers is not in the document]
Comment (rezaansyed) [reply packages/app/src/cli/prompts/config.ts]: hmm that's weird. The config I had was from a newly generated app

File: packages/app/src/cli/prompts/config.ts [hunk 2]
```diff
@@ -63,3 +71,32 @@ export function validate(value: string): string | undefined {
...
   if (result.length > 238) return 'The file name is too long.'
 }
+
+export async function confirmPushChanges(options: PushOptions, app: App) {
+  if (options.force) return true
+
+  const {configuration, configurationPath} = options
+  const localConfigurationToml = await readFile(configurationPath)
+  const localConfiguration = decodeToml(localConfigurationToml)
+  const remoteConfiguration = mergeAppConfiguration({configuration} as AppInterface, app as OrganizationApp)
```
Comment (RyanDJLee) [inline packages/app/src/cli/prompts/config.ts:81]: This should just be the current configuration values on the platform without the merge.

**Review by RyanDJLee — CHANGES_REQUESTED** [review]
Looks great! But we may need to change to logic for what we want to compare.

File: packages/app/src/cli/prompts/config.ts (hunk 2 above)
Comment (rezaansyed) [reply packages/app/src/cli/prompts/config.ts:81]: I'm noticing that the diff captures the changes without the merge so I think it's fine:
image

**Review by rezaansyed — APPROVED** [review]
Tophatted and the changes look good to me! 🎉

File: packages/app/src/cli/prompts/config.ts (hunk 2 above)
Comment (gonzaloriestra) [reply packages/app/src/cli/prompts/config.ts:81]: If we compare against the raw platform configuration, then some local values would always appear in the diff (like `automatically_update_urls_on_dev`).

Also, we just have the app here, so we would need to create a configuration object to be able to compare, selecting the desired fields. Which is basically what `mergeAppConfiguration` does. Using the same logic than in `link` ensures we will never find changes when we push right after linking.

**Comment by gonzaloriestra** [comment]
I've made a couple of last-minute changes:
- The app URL returns always with and ending `/` (`https://example.com/`) even if it's pushed without it (`https://example.com`), and the cloudflare URLs we are setting with dev don't have it. So I'm removing it from the remote fields.
- The link command was removing some local configuration (like `automatically_update_urls_on_dev`). I changed it to preserve it and only overwrite the remote fields.

**Review by RyanDJLee — APPROVED** [review]
Looks good!

**Pull request merged**